File: cart/src/main/kotlin/com/hateoasatscale/cart/controllers/cart/Product.kt
package com.hateoasatscale.cart.controllers.cart

import com.fasterxml.jackson.annotation.JsonCreator
import org.springframework.hateoas.Link
import org.springframework.hateoas.RepresentationModel
import java.math.BigDecimal

class Product @JsonCreator constructor(
    private val usersServiceUrl: String,
    private val id: Long,
    val name: String,
    val reference: String,
    val price: BigDecimal,
) : RepresentationModel<Product>() {

    init {
        addSelfLink()
    }

    private fun addSelfLink() {
        add(Link.of("$usersServiceUrl/products/$id"))
    }
}

File: client/src/app/shared/product.ts
import {HateoasModel} from './hateoas.model';

export class Product extends HateoasModel {
  constructor(
    readonly name: string,
    readonly reference: string,
    readonly price: number,
  ) {
    super();
  }

  static from(json: any) {
    let product = new Product(json.name, json.reference, json.price);
    product._links = json._links;
    return product;
  }
}
